### PR TITLE
map Interop.Sys.LockType to distro-specific values

### LINKED ISSUE #26726: System.IO.FileSystem tests failing on FreeBSD
```
 /usr/home/furt/git/source-build3/src/corefx/bin/tests/System.IO.FileSystem.Tests/netcoreapp-FreeBSD-Debug-x64 /usr/home/furt/git/source-build3/src/corefx/src/System.IO.FileSystem/tests
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.

  Discovering: System.IO.FileSystem.Tests
  Discovered:  System.IO.FileSystem.Tests
  Starting:    System.IO.FileSystem.Tests

  Unhandled Exception: Xunit.Sdk.ThrowsException: Assert.Throws() Failure
  Expected: typeof(System.IO.IOException)
  Actual:   (No exception was thrown)
     at Xunit.Assert.Throws(Type exceptionType, Exception exception)
     at Xunit.Assert.Throws[T](Action testCode)
     at System.IO.Tests.FileStream_LockUnlock.<>c.<OverlappingRegionsFromOtherProcess_ThrowsException>b__7_0(String secondPath, String secondPos, String secondLen) in /usr/home/furt/git/source-build3/src/corefx/src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs:line 178
  --- End of stack trace from previous location where exception was thrown ---
     at RemoteExecutorConsoleApp.Program.Main(String[] args) in /usr/home/furt/git/source-build3/src/corefx/src/Common/tests/System/Diagnostics/RemoteExecutorConsoleApp/RemoteExecutorConsoleApp.cs:line 92
  SystemNative_LockFileRegion on 36 with 0x1
     System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 0, firstLength: 10, secondPosition: 1, secondLength: 2) [FAIL]
        System.Diagnostics.RemoteExecutorTestBase+RemoteInvokeHandle+RemoteExecutionException : Remote process failed with an unhandled exception.
        Stack Trace:

           Child exception:
             Xunit.Sdk.ThrowsException: Assert.Throws() Failure
           Expected: typeof(System.IO.IOException)
           Actual:   (No exception was thrown)
           /usr/home/furt/git/source-build3/src/corefx/src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs(178,0): at System.IO.Tests.FileStream_LockUnlock.<>c.<OverlappingRegionsFromOtherProcess_ThrowsException>b__7_0(String secondPath, String secondPos, String secondLen)

           Child process:
             System.IO.FileSystem.Tests, Version=4.1.1.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb System.IO.Tests.FileStream_LockUnlock+<>c Int32 <OverlappingRegionsFromOtherProcess_ThrowsException>b__7_0(System.String, System.String, System.String)

           Child arguments:
             /tmp/FileStream_LockUnlock_u2fx1sfl.4r0/OverlappingRegionsFromOtherProcess_ThrowsException_167_e7a3c8e1, 1, 2


  Unhandled Exception: Xunit.Sdk.ThrowsException: Assert.Throws() Failure
  Expected: typeof(System.IO.IOException)
  Actual:   (No exception was thrown)
     at Xunit.Assert.Throws(Type exceptionType, Exception exception)
     at Xunit.Assert.Throws[T](Action testCode)
     at System.IO.Tests.FileStream_LockUnlock.<>c.<OverlappingRegionsFromOtherProcess_ThrowsException>b__7_0(String secondPath, String secondPos, String secondLen) in /usr/home/furt/git/source-build3/src/corefx/src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs:line 178
  --- End of stack trace from previous location where exception was thrown ---
     at RemoteExecutorConsoleApp.Program.Main(String[] args) in /usr/home/furt/git/source-build3/src/corefx/src/Common/tests/System/Diagnostics/RemoteExecutorConsoleApp/RemoteExecutorConsoleApp.cs:line 92
  SystemNative_LockFileRegion on 36 with 0x1
     System.IO.Tests.FileStream_LockUnlock.OverlappingRegionsFromOtherProcess_ThrowsException(fileLength: 10, firstPosition: 3, firstLength: 5, secondPosition: 3, secondLength: 5) [FAIL]
        System.Diagnostics.RemoteExecutorTestBase+RemoteInvokeHandle+RemoteExecutionException : Remote process failed with an unhandled exception.
        Stack Trace:

           Child exception:
             Xunit.Sdk.ThrowsException: Assert.Throws() Failure
           Expected: typeof(System.IO.IOException)
           Actual:   (No exception was thrown)
           /usr/home/furt/git/source-build3/src/corefx/src/System.IO.FileSystem/tests/FileStream/LockUnlock.cs(178,0): at System.IO.Tests.FileStream_LockUnlock.<>c.<OverlappingRegionsFromOtherProcess_ThrowsException>b__7_0(String secondPath, String secondPos, String secondLen)

           Child process:
             System.IO.FileSystem.Tests, Version=4.1.1.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb System.IO.Tests.FileStream_LockUnlock+<>c Int32 <OverlappingRegionsFromOtherProcess_ThrowsException>b__7_0(System.String, System.String, System.String)

           Child arguments:
             /tmp/FileStream_LockUnlock_gog4cn0u.lvd/OverlappingRegionsFromOtherProcess_ThrowsException_167_07dedcd6, 3, 5

     System.IO.Tests.FileInfo_Exists.LockedFileExists [FAIL]

```

All the failing tests are marked with `[PlatformSpecific(~TestPlatforms.OSX)]` so there is possibility FreeBSD behaves as OSX.
This should be investigated.

### PR DESCRIPTION
**Pull request by adamsitnik**
https://github.com/torvalds/linux/blob/fcadab740480e0e0e9fa9bd272acd409884d431a/arch/alpha/include/uapi/asm/fcntl.h#L48-L50
https://github.com/freebsd/freebsd-src/blob/fb8c2f743ab695f6004650b58bf96972e2535b20/sys/sys/fcntl.h#L277-L279

fixes #26726

### CODE REVIEW

File: src/libraries/Native/Unix/System.Native/pal_io.c [hunk 1]
```diff
@@ -1263,9 +1263,30 @@ char* SystemNative_RealPath(const char* path)
...
+    {
+        case 0:
+            return F_RDLCK;
+        case 1:
+            return F_WRLCK;
+        case 2:
+            return F_UNLCK;
+        default:
+            assert_msg(false, "Unknown Lock Type", (int)managedLockType);
+            return -1;
```
Comment (stephentoub) [inline src/libraries/Native/Unix/System.Native/pal_io.c:1282]: What's to guarantee one of these values isn't actually -1?

Given that this is an internal implementation detail (we own all call sites calling this and only pass in known const values), I suggest we just make this:
```C++
case 0:
    return F_RDLCK;
case 1:
    return F_WRLCK;
default:
    assert_msg(managedLockType == 2, "...");
    return F_UNLCK;
```
or something like that.  And then undo the unixLockType < 0 change at the call site to this helper.

**Review by carlossanlop — APPROVED** [review]
LGTM, pending adding Stephen's suggestion.

File: src/libraries/Native/Unix/System.Native/pal_io.c [hunk 2]
```diff
@@ -1263,9 +1263,30 @@ char* SystemNative_RealPath(const char* path)
...
+    }
```
Comment (adamsitnik) [inline src/libraries/Native/Unix/System.Native/pal_io.c:1283]:
```suggestion
        default:
            assert_msg(managedLockType == 2, "Unknown Lock Type");
            return F_UNLCK;
    }
```

File: src/libraries/Native/Unix/System.Native/pal_io.c [hunk 3]
```diff
@@ -1263,9 +1263,30 @@ char* SystemNative_RealPath(const char* path)
...
+}
+
 int32_t SystemNative_LockFileRegion(intptr_t fd, int64_t offset, int64_t length, int16_t lockType)
 {
-    if (offset < 0 || length < 0)
+    int16_t unixLockType = ConvertLockType(lockType);
+    if (offset < 0 || length < 0 || unixLockType < 0)
```
Comment (adamsitnik) [inline src/libraries/Native/Unix/System.Native/pal_io.c:1289]:
```suggestion
    if (offset < 0 || length < 0)
```

File: src/libraries/Native/Unix/System.Native/pal_io.c [hunk 4]
```diff
@@ -1263,9 +1263,30 @@ char* SystemNative_RealPath(const char* path)
...
+    switch (managedLockType)
+    {
+        case 0:
+            return F_RDLCK;
+        case 1:
+            return F_WRLCK;
+        case 2:
+            return F_UNLCK;
+        default:
+            assert_msg(false, "Unknown Lock Type", (int)managedLockType);
```
Comment (adamsitnik) [inline src/libraries/Native/Unix/System.Native/pal_io.c]:
```suggestion
            assert_msg(managedLockType == 2, "Unknown Lock Type", (int)managedLockType);
```

File: src/libraries/Native/Unix/System.Native/pal_io.c [hunk 5]
```diff
@@ -1263,8 +1263,27 @@ char* SystemNative_RealPath(const char* path)
...
+    // https://github.com/torvalds/linux/blob/fcadab740480e0e0e9fa9bd272acd409884d431a/arch/alpha/include/uapi/asm/fcntl.h#L48-L50
+    // https://github.com/freebsd/freebsd-src/blob/fb8c2f743ab695f6004650b58bf96972e2535b20/sys/sys/fcntl.h#L277-L279
+    switch (managedLockType)
+    {
+        case 0:
+            return F_RDLCK;
+        case 1:
+            return F_WRLCK;
+        default:
+            assert_msg(managedLockType == 2, "Unknown Lock Type");
```
Comment (adamsitnik) [inline src/libraries/Native/Unix/System.Native/pal_io.c:1279]:
```suggestion
            assert_msg(managedLockType == 2, "Unknown Lock Type", (int)managedLockType);
```

**Comment by danmoseley** [comment]
I wonder if it would be less confusing if we changed
```
        internal enum LockType : short
        {
            F_RDLCK = 0,    // shared or read lock
            F_WRLCK = 1,    // exclusive or write lock
            F_UNLCK = 2     // unlock
        }
```
to
```
        internal enum LockType : short
        {
            SharedOrReadLock = 0,  
            ExclusiveOrWriteLock = 1,   
            Unlock = 2   
        }
```
but I don't have a strong opinion.

**Comment by adamsitnik** [comment]
> I wonder if it would be less confusing if we changed

The managed enum is used in only [one place](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.Lock.Unix.cs) in the code and it's unlikely that we are going to reuse it in other places, so I am going to keep it as it is.

**Pull request merged**